Codechange: Increase scrollbar length limit to 32bits

LordAro opened the pull request:
Also makes scrollbar internals properly use unsigned numbers. Bumped scrollbar limit to uint(32) as well, because why not

Fixes #7613

stormcone commented:
There are 3 occurrences of `this->group_sb->SetCount((uint)this->groups.size());` in `group_gui.cpp`, where I think the `(uint)` could be removed.

LordAro commented:
good catch, seems I'd missed quite a few

glx22 commented on src/newgrf_debug_gui.cpp:
```diff
@@ -402,8 +402,7 @@ struct NewGRFInspectWindow : Window {
-		if (offset < 0 || offset >= this->vscroll->GetCapacity()) return;
+		if (offset + this->vscroll->GetPosition() >= this->vscroll->GetCapacity()) return;
 
 		::DrawString(r.left + LEFT_OFFSET, r.right - RIGHT_OFFSET, r.top + TOP_OFFSET + (offset * this->resize.step_height), buf, TC_BLACK);
```
hmm `offset` passed to `DrawString()` is off by `this->vscroll->GetPosition()` I think

glx22 commented on src/newgrf_gui.cpp:
```diff
@@ -1525,10 +1528,10 @@ struct NewGRFWindow : public Window, NewGRFScanCallback {
+		if (new_pos >= this->avails.size()) new_pos = this->avail_pos - 1;
 		this->avail_pos = new_pos;
-		if (new_pos >= 0) this->avail_sel = this->avails[new_pos];
+		this->avail_sel = this->avails[new_pos];
```
I think a safety check is still needed, `new_pos` can be `UINT_MAX` if `this->avail_pos` was 0 and only 1 element in `this->avails`

TrueBrain commented on src/group_gui.cpp:
```diff
@@ -502,8 +502,8 @@ class VehicleGroupWindow : public BaseVehicleListWindow {
-		this->group_sb->SetCount(static_cast<int>(this->groups.size()));
-		this->vscroll->SetCount(static_cast<int>(this->vehgroups.size()));
+		this->group_sb->SetCount(this->groups.size());
+		this->vscroll->SetCount(this->vehicles.size());
```
Rebase mistake I think :) `vehicles` should be `vehgroups`?

TrueBrain commented on src/newgrf_debug_gui.cpp:
```diff
@@ -402,8 +402,8 @@ struct NewGRFInspectWindow : Window {
 		vseprintf(buf, lastof(buf), format, va);
 		va_end(va);
 
+		if (offset + this->vscroll->GetPosition() >= this->vscroll->GetCapacity()) return;
```
Are you sure this is correct?

In the old code, `offset` seems to go from 0..N, with `this->vscroll` could have a Position of M, and a Capacity of X. (where M and X have no relation to each other).

It returns if `offset - M < 0` or `offset - M >= X`. In other words, if `offset < M` or `offset >= X + M`. Basically, if the offset is within the range of the scrollbar.

Now in the new code you do the upper-bound check, but I do not see the lower-bound check. But the statement that is here, seems to be wrong (`offset + M >= X`). It mostly triggered me as I couldn't make sense of this if-statement ;) To me it reads the old code does: if not `M <= offset < M + X`, return. Or in real C++ code:

```
/* Ensure the offset is within the vscroll range */
if (offset < this->vscroll->GetPosition()) retun;
if (offset >= this->vscroll->GetPosition() + this->vscroll->GetCapacity()) return;
```

I might be completely wrong, so please do your own analysis again too, but your new line of code simply does not make sense to me: `offset + M >= X` .. like .. if you scroll for X, this is always true. So nothing renders anymore?

TrueBrain commented on src/widget.cpp:
```diff
@@ -1636,30 +1636,30 @@ NWidgetCore *NWidgetMatrix::GetWidgetFromPos(int x, int y)
 
 			child->AssignSizePosition(ST_RESIZE, offs_x, offs_y, child->smallest_x, child->smallest_y, rtl);
-			child->SetLowered(this->clicked == sub_wid);
+			child->SetLowered(this->clicked == (int)sub_wid);
```
You happen to know why `this->clicked` still needs to be signed? (did not check the code, it was easier to ask you :D)

LordAro replied:
-1 is used as "nothing clicked". Technically speaking I suppose I could convert it to use UINT_MAX as none instead, but... eh

TrueBrain commented on src/network/network_content_gui.cpp:
```diff
@@ -810,7 +810,9 @@ class NetworkContentListWindow : public Window, ContentCallback {
 					this->content.ToggleSortOrder();
-					if (this->content.size() > 0) this->list_pos = (int)this->content.size() - this->list_pos - 1;
+					if (this->content.size() > 0) {
+						this->list_pos = (this->list_pos - 1 > this->content.size()) ? 0 : this->content.size() - this->list_pos - 1;
```
shouldn't this be `+ 1`, as in:

`this->list_pos = (this->list_pos + 1 > this->content.size()) ? 0 : this->content.size() - this->list_pos - 1;`

?

TrueBrain commented on src/ai/ai_gui.cpp:
```diff
@@ -1176,7 +1176,7 @@ struct AIDebugWindow : public Window {
 		}
 		if (this->autoscroll) {
-			int scroll_pos = max(0, log->used - this->vscroll->GetCapacity());
+			uint scroll_pos = (this->vscroll->GetCapacity() > log->used ? 0 : log->used - this->vscroll->GetCapacity());
```
In general, and that goes for all the next instances here too, I have a hard time processing the direction of these statements. I would expect:

`uint scroll_pos = (log->used < this->vscroll->GetCapacity() ? 0 : log->used - this->vscroll->GetCapacity());`

As than both statements read in the same order. This flipping around requires more processing power in my brain. But I fully admit that this can be a local problem (as in: in my brain problem), but maybe something to consider :)

(and yes, the inverse of your statement is `<=`, but in this case that totally doesn't matter .. I guess for both our cases, adding the `=` would be a slight reduction of CPU time :P)

TrueBrain replied:
No, I take this back; I write both forms when I am writing code myself. Seems the order just depends on how/when I am reading it. This is really weird, how brains work :D

TrueBrain commented on src/widget_type.h:
```diff
@@ -665,14 +665,14 @@ class Scrollbar {
 		this->count = num;
-		num -= this->cap;
-		if (num < 0) num = 0;
+		if (this->cap > num) {
```
To come back to `>` vs `<`, this was the one that triggered me most. Here:

```
if (num < this->cap) {
  num = 0;
} else {
  num -= this->cap;
}
```

Reads much more natural to me. But again, this can completely be because my brain is primed like that, I am just mentioning it so you can consider it :)

TrueBrain commented on src/company_gui.cpp:
```diff
@@ -731,7 +731,8 @@ struct SelectCompanyLiveryWindow : public Window {
 		for (uint i = 0; i < this->rows; i++) {
 			if (this->groups[i]->index == sel) {
-				this->vscroll->SetPosition(Clamp(i - this->vscroll->GetCapacity() / 2, 0, std::max(this->vscroll->GetCount() - this->vscroll->GetCapacity(), 0)));
+				uint scroll_pos = UnderflowSafeSub(this->vscroll->GetCapacity(), this->vscroll->GetCount());
```
This used to be `Count - Capactity`, and now seems flipped?

stormcone commented on src/core/math_func.hpp:
```diff
@@ -159,6 +159,18 @@ static inline uint16 ClampToU16(const uint64 a)
+static inline uint UnderflowSafeSub(uint a, uint b)
+{
+	if (b < a) return 0;
+	return a - b;
```
Shouldn't it be `if (b > a)`?

TrueBrain replied:
He is testing us, I am telling you :) Just to see if the reviewers are awake .. I clearly wasn't :P

TrueBrain reviewed (dismissed):
No longer crashes the game, tried a few windows, scrolling seems to work fine.

I am just there will be some window that doesn't work as we want or what-ever, but we will fix that when people report it :D

LordAro commented:
Shouldn't be merged yet, lots of size_t -> uint warnings on Windows. Going to have to readd a load of casts :(

LordAro committed:
Codechange: Use 32bit integers for scrollbars

LordAro commented:
I've updated it to just increase the size, no fiddling around with type-signedness.

...Except for some reason changing it from `unsigned short` to `unsigned int` has caused the compiler to no longer be able to make as many coersions as it was able to previously. I've fixed them in a "minimal" way. A better solution would be to go through the whole system to check everything for over/underflows, as I've no doubt there are some lurking somewhere.

nielsmh commented on src/company_gui.cpp:
```diff
@@ -731,7 +731,7 @@ struct SelectCompanyLiveryWindow : public Window {
 		for (uint i = 0; i < this->rows; i++) {
 			if (this->groups[i]->index == sel) {
-				this->vscroll->SetPosition(Clamp(i - this->vscroll->GetCapacity() / 2, 0, std::max(this->vscroll->GetCount() - this->vscroll->GetCapacity(), 0)));
+				this->vscroll->SetPosition(Clamp(i - this->vscroll->GetCapacity() / 2, 0, std::max<int>(this->vscroll->GetCount() - this->vscroll->GetCapacity(), 0)));
```
This line is really confusing. Actually I'm not sure why `Scrollbar::SetPosition` assert-checks that the parameter is in range, instead of just clamping it into range itself. Letting the scrollbar handle that would make all the using code simpler, wouldn't it?

nielsmh reviewed (changes requested):
This code has a lot of typecasting to stop the compiler complaining about signed/unsigned type ranges, but it's not actually range safe as far as I can tell. If the scrollbar isn't going to work properly with more than 2^31 elements in the list anyway, just make it use a plain signed int type and do some sanity checks against negative numbers in a few strategic locations.
